pkg/node/agent.go

```diff
@@ -0,0 +1,248 @@
+package node
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/coreos/bootkube/pkg/atomic"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_3"
+)
+
+// Agent reacts to annotation updates on the node it is running on.
+// When the annotation regarding the desired config is updated,
+// the Agent will update the configuration on disk and restart the
+// on-host kubelet.
+type Agent struct {
+	// NodeName is the name of the node the agent is running on.
+	NodeName string
+	// Client is an API server client.
+	Client clientset.Interface
+	// SysdConn is a sytemd dbus connection, used for restarting the on-host
+	// kubelet via systemd.
+	SysdConn dbusConn
+	// ConfigMapStore is updated via an informer.
+	ConfigMapStore cache.Store
+}
+
+type dbusConn interface {
+	Reload() error
+	RestartUnit(string, string, chan<- string) (int, error)
+}
+
+const (
+	desiredConfigAnnotation = "node-agent.alpha.coreos.com/desired-config"
+	currentConfigAnnotation = "node-agent.alpha.coreos.com/current-config"
+	kubeletVersionKey       = "KUBELET_VERSION"
+	kubeletConfigKey        = "KUBELET_CONFIG"
+	kubeletFlagsKey         = "KUBELET_OPTS"
+	defaultKubeletEnvPath   = "/etc/kubernetes/kubelet.env"
+	kubeletService          = "kubelet.service"
+	configMapFlagsKey       = "kubelet-flags"
+)
+
+// NodeUpdateCallback is called via an informer watching Node objects.
+// When called, it will check the Node annotations looking for a new desired
+// config. When it finds a new desired config, it will update the on-disk config
+// and restart the on-host kubelet.
+func (a *Agent) NodeUpdateCallback(_, newObj interface{}) {
+	glog.Infof("begin node update callback")
+
+	node, ok := newObj.(*v1.Node)
+	if !ok {
+		glog.Errorf("received unexpected type: %T (expected *v1.Node)", newObj)
+		return
+	}
+
+	// Check annotations on Node.
+	if node.Annotations == nil {
+		node.Annotations = make(map[string]string)
+	}
+
+	var desiredConfig map[string]string
+
+	onDiskConfig, err := parseKubeletEnvFile(defaultKubeletEnvPath)
+	if err != nil {
+		glog.Error(err)
+		goto update
+	}
+
+	// Obtain and validate the desired configuration.
+	desiredConfig, err = a.getDesiredConfig(node)
+	if err != nil {
+		glog.Error(err)
+	}
+
+update:
```

Comment:
Also it looks like we would still continue past the desired config even if we can't parse it anyway, so this essentially is just skipping one possible log entry.